### PR TITLE
Victor PP

### PR DESCRIPTION
**Pull request by JamesWatling**

### CODE REVIEW

**Comment by JamesWatling** [comment]
@guavapass/dev-awesome 👀

File: app/models/booking.rb [hunk 1]
```diff
@@ -1,3 +1,38 @@
...
+
+  def no_show!
+    check_no_show_condition!
+    ActiveRecord::Base.transaction do
+      self.update(status: "no_show")
+
+      # we should be aware of transaction issue here
+      if Time.now - lesson.starts_at <= NO_SHOW_WINDOW
+        # we raise an error here if the record is not actually persisted
+        NoShowCharge.create!()
```
Comment (beirigo) [inline app/models/booking.rb:14]: this charge doesn't have a booking reference, it's useless

File: app/models/booking.rb [hunk 2]
```diff
@@ -1,3 +1,38 @@
 class Booking < ApplicationRecord
+  belongs_to :lesson
```
Comment (beirigo) [inline app/models/booking.rb:2]: the NoShowCharge and BillingAdjustment associations are missing.

File: app/models/booking.rb [hunk 3]
```diff
@@ -1,3 +1,38 @@
...
+  private
+  def check_no_show_condition!
+    if self.status == "cancelled"
+      raise StandardError, "Cannot mark no show for `cancelled` bookings."
+    elsif self.status == "no_show"
+      raise StandardError, "Booking already marked as `no_show`."
+    end
+  end
+
+  def previous_payout_date
```
Comment (beirigo) [inline app/models/booking.rb:34]: IF you run this before the 5th, it will give you the next payout date, not the previous

**Review by beirigo — COMMENTED** [review]
nightmare

File: app/models/booking.rb (hunk 2 above)
Comment (JamesWatling) [reply app/models/booking.rb:2]: Yep

File: app/models/booking.rb (hunk 3 above)
Comment (JamesWatling) [reply app/models/booking.rb:34]: Yep

**Pull request closed without merging**